c_src/odbcserver.c

```diff
@@ -3012,3 +2843,93 @@ static void str_tolower(char *str, int len)
+   SQL_SUCCESS_WITH_INFO nd the SQLSTATE will have the value 01004 (Data
+   truncated). The application can then use the same column number to
+   retrieve subsequent parts of the data until SQLGetData returns
+   SQL_SUCCESS, indicating that all data for the column has been retrieved.
+*/
+/* Allocs and sets bufferptr
+ * Sets result_len
+ */
+void retrive_long_data(
+        db_column column,
+        int column_nr,
+        SQLSMALLINT target_type,
+        char** result_buf,
+        int* result_len,
+        int sizeof_element,
+        int term_bytes,
+        db_state* state) {
+    // That moment, when IBM has better docs than MS
+    // https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_74/cli/rzadpfnbndpm.htm
+    SQLRETURN     rc;
+    const int chunk_bytes = sizeof_element * 5001; // Should be odd to avoid different alignment issues
+    int buff_bytes = chunk_bytes;
+    int offset_bytes = 0;
+    SQLLEN byte_len_or_ind = 0;
+    int truncated;
+    int got_bytes;
+    diagnos diagnos;
+
+    // pointer to the beginning of the string
+    char* bufferptr = (void*) safe_malloc(buff_bytes);
+
+    // Retrieve data in parts.
+    // Sets byte_len_or_ind
+    // Writes chunk_bytes bytes into a buffer. Though 2 last bytes are a null terminator.
+    // So, it's (chunk_bytes-2) actual bytes
+    while (1) {
+        // Free space for a new chunk in bytes.
+        // This condition applies:
+        // int chunk_bytes = buff_bytes - offset_bytes;
+
+        /*
+         Writes maximum BufferLength bytes into a TargetValuePtr
+         (BufferLength - 2) bytes of actual data + 2 bytes as a null terminator
+    
+         SQLRETURN SQLGetData(
+          SQLHSTMT       StatementHandle,
+          SQLUSMALLINT   Col_or_Param_Num,
+          SQLSMALLINT    TargetType,
+          SQLPOINTER     TargetValuePtr,
+          SQLLEN         BufferLength,
+          SQLLEN *       StrLen_or_IndPtr); 
+         */
+        byte_len_or_ind = 0;
+        rc = SQLGetData(statement_handle(state),
+                        (SQLSMALLINT)(column_nr+1),
+                        target_type,
+                        bufferptr + offset_bytes, // where to append data
+                        chunk_bytes,
+                        &byte_len_or_ind);
+        if (byte_len_or_ind == SQL_NULL_DATA) {
+            *result_buf = NULL;
+            *result_len = 0;
+            return;
+        }
+//      if (rc == SQL_NO_DATA) break;
+        if (!sql_success(rc)) break;
+
+        diagnos = get_diagnos(SQL_HANDLE_STMT, statement_handle(state), extended_errors(state));
+        truncated = (strcmp((char *)diagnos.sqlState, TRUNCATED) == 0);
+
+        // Number of written bytes without NULL terminator
+        got_bytes = (byte_len_or_ind >= chunk_bytes) ? (chunk_bytes - term_bytes) : byte_len_or_ind;
+        if (byte_len_or_ind == SQL_NULL_DATA) got_bytes = 0;
+        offset_bytes += got_bytes;
```

Comment:
This is confusing to quick-read, I'd do something like
```suggestion
        if (byte_len_or_ind == SQL_NULL_DATA)
            got_bytes = 0;
        offset_bytes += got_bytes;
```
or
```suggestion
        if (byte_len_or_ind == SQL_NULL_DATA) got_bytes = 0;
        
        offset_bytes += got_bytes;
```